lib/hl7/datetime_components.rb

```diff
@@ -0,0 +1,125 @@
+require 'active_support/core_ext/time'
+
+module HL7
+  class DatetimeComponents
+    attr_reader :year, :month, :day, :hour, :minute, :second, :fraction, :offset_seconds
+
+    def initialize(year, month = nil, day = nil, hour = nil, minute = nil, second = nil, fraction = nil, offset_seconds: nil)
+      @year = year
+      @month = month
+      @day = day
+      @hour = hour
+      @minute = minute
+      @second = second
+      @fraction = fraction
+      @offset_seconds = offset_seconds
+      validate!
+    end
+
+    def to_time(zone:)
+      full_seconds = second.to_i + fraction.to_f
+
+      if offset_seconds.nil?
+        Time.use_zone(zone) do
+          Time.zone.local(year, month, day, hour, minute, full_seconds)
+        end
+      else
+        Time.new(year, month, day, hour, minute, full_seconds, offset_seconds).
+          in_time_zone(zone)
+      end
+    end
+
+    def ==(other)
+      other.is_a?(DatetimeComponents) &&
+        self.to_time(zone: "UTC").eql?(other.to_time(zone: "UTC"))
+    end
+
+    private
+
+    def validate!
+      # year is not allowed to be nil
+      AttrSpec.new(:year, -9999..9999).validate_non_nil!(year)
```

Comment:
You're going to regret this in ~8,000 years.

Reply:
As it happens, this is both unnecessary and overly liberal. Unnecessary, because we just matched `\d{4}`, and overly liberal because there's no place in that pattern for a minus sign. /shrug.